src/main/scala/cromwell/engine/backend/jes/JesBackend.scala

```diff
@@ -0,0 +1,199 @@
+  )
+
+  sealed trait JesParameter { // FIXME: Perhaps not the best name
+    def name: String
+    def gcs: String
+    def local: Path
+
+    final def isInput = this.isInstanceOf[JesInput]
+    final def isOutput = !isInput
+    // FIXME: Perhaps not the best name
+    final def toGoogleParameter = new Parameter().setName(name).setValue(local.toString).setType("REFERENCE")
+  }
+
+  final case class JesInput(name: String, gcs: String, local: Path) extends JesParameter
+  final case class JesOutput(name: String, gcs: String, local: Path) extends JesParameter
+}
+
+class JesBackend extends Backend with LazyLogging {
+
+  /**
+   * Takes a path in GCS and comes up with a local path which is unique for the given GCS path
+   * @param gcsPath The input path
+   * @return A path which is unique per input path
+   */
+  def localFilePathFromCloudStoragePath(gcsPath: GoogleCloudStoragePath): Path = {
+    Paths.get("/cromwell_root/" + gcsPath.bucket + "/" + gcsPath.objectName)
+  }
+
+  /**
+   * Takes a single FQN to WdlValue pair and maps google cloud storage (GCS) paths into an appropriate local file path.
+   * If the input is not a WdlFile, or the WdlFile is not a GCS path, the mapping is a noop.
+   *
+   * @param fqn the FQN of the input variable
+   * @param wdlValue the value of the input
+   * @return a new FQN to WdlValue pair, with WdlFile paths modified if appropriate.
+   */
+  def mapInputValue(fqn: String, wdlValue: WdlValue): (String, WdlValue) = {
+    wdlValue match {
+      case WdlFile(path) =>
+        GoogleCloudStoragePath.parse(path) match {
+          case Success(gcsPath) => (fqn, WdlFile(localFilePathFromCloudStoragePath(gcsPath).toString))
+          case Failure(e) => (fqn, wdlValue)
+        }
+      case _ => (fqn, wdlValue)
+    }
+  }
+
+  override def adjustInputPaths(call: Call, inputs: CallInputs): CallInputs = {
+    inputs map { case (k,v) => mapInputValue(k,v) }
+  }
+
+  override def adjustOutputPaths(call: Call, outputs: CallOutputs): CallOutputs = outputs
+  
+  // No need to copy GCS inputs for the workflow we should be able to directly reference them
+  override def initializeForWorkflow(workflow: WorkflowDescriptor): HostInputs = workflow.actualInputs
+
+  // FIXME: signature is weird
+  def localizeTaskOutput(taskOutput: TaskOutput, callGcsPath: String, scopedLookupFunction: ScopedLookupFunction, engineFunctions: JesEngineFunctions): Try[Option[JesOutput]] = {
+    taskOutput.wdlType match {
+      case WdlFileType =>
+        taskOutput.expression.evaluate(scopedLookupFunction, engineFunctions) match {
+          case Success(v) => Success(Option(JesOutput(taskOutput.name, s"$callGcsPath/${taskOutput.name}", Paths.get(v.valueString))))
+          case Failure(e) => Failure(new IllegalArgumentException(s"JES requires File outputs to be determined prior to running, but ${taskOutput.name} can not."))
+        }
+      case _ => Success(None)
+    }
+  }
+
+  // FIXME: Signature is weird
```

Comment:
I was more complaining about the inputs but I don't remember my exact complaint except that I thought it was messy